### DISCUSSION TITLE
datasets/pose/hand-keypoints/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#16544] [Docs]
# datasets/pose/hand-keypoints/

Explore the hand keypoints estimation dataset for advanced pose estimation. Learn about datasets, pretrained models, metrics, and applications for training with YOLO.

https://docs.ultralytics.com/datasets/pose/hand-keypoints/

**Comment by redsuning** [comment]
train: WARNING ⚠️ G:\datasets\hand\hand-keypoints\train\images\IMG_00001070.jpg: ignoring corrupt image/label: non-normalized or out of bounds coordinates [     1.0379]
train: WARNING ⚠️ G:\datasets\hand\hand-keypoints\train\images\IMG_00001796.jpg: ignoring corrupt image/label: non-normalized or out of bounds coordinates [     1.0006]
train: WARNING ⚠️ G:\datasets\hand\hand-keypoints\train\images\IMG_00001800.jpg: ignoring corrupt image/label: non-normalized or out of bounds coordinates [     1.0027]
train: WARNING ⚠️ G:\datasets\hand\hand-keypoints\train\images\IMG_00001801.jpg: ignoring corrupt image/label: non-normalized or out of bounds coordinates [     1.0049]
train: WARNING ⚠️ G:\datasets\hand\hand-keypoints\train\images\IMG_00001886.jpg: ignoring corrupt image/label: non-normalized or out of bounds coordinates [     1.0051      1.0038      1.0088      1.0114]
train: WARNING ⚠️ G:\datasets\hand\hand-keypoints\train\images\IMG_00001887.jpg: ignoring corrupt image/label: non-normalized or out of bounds coordinates [     1.0048      1.0062      1.0099      1.0121]
train: WARNING ⚠️ G:\datasets\hand\hand-keypoints\train\images\IMG_00001888.jpg: ignoring corrupt image/label: non-normalized or out of bounds coordinates [          1      1.0017]
train: WARNING ⚠️ G:\datasets\hand\hand-keypoints\train\images\IMG_00001994.jpg: ignoring corrupt image/label: non-normalized or out of bounds coordinates [     1.0059]
train: WARNING ⚠️ G:\datasets\hand\hand-keypoints\train\images\IMG_00001998.jpg: ignoring corrupt image/label: non-normalized or out of bounds coordinates [     1.0044]
train: WARNING ⚠️ G:\datasets\hand\hand-keypoints\train\images\IMG_00001999.jpg: ignoring corrupt image/label: non-normalized or out of bounds coordinates [      1.007]
train: WARNING ⚠️ G:\datasets\hand\hand-keypoints\train\images\IMG_00002000.jpg: ignoring corrupt image/label: non-normalized or out of bounds coordinates [     1.0083]
train: WARNING ⚠️ G:\datasets\hand\hand-keypoints\train\images\IMG_00002001.jpg: ignoring corrupt image/label: non-normalized or out of bounds coordinates [     1.0146]
train: WARNING ⚠️ G:\datasets\hand\hand-keypoints\train\images\IMG_00002002.jpg: ignoring corrupt image/label: non-normalized or out of bounds coordinates [     1.0178]
train: WARNING ⚠️ G:\datasets\hand\hand-keypoints\train\images\IMG_00002003.jpg: ignoring corrupt image/label: non-normalized or out of bounds coordinates [     1.0178]
train: WARNING ⚠️ G:\datasets\hand\hand-keypoints\train\images\IMG_00002143.jpg: ignoring corrupt image/label: non-normalized or out of bounds coordinates [     1.0582      1.0216]
train: WARNING ⚠️ G:\datasets\hand\hand-keypoints\train\images\IMG_00002144.jpg: ignoring corrupt image/label: non-normalized or out of bounds coordinates [     1.0641      1.0333]
train: WARNING ⚠️ G:\datasets\hand\hand-keypoints\train\images\IMG_00002202.jpg: ignoring corrupt image/label: non-normalized or out of bounds coordinates [     1.0041      1.0052       1.008      1.0093      1.0108      1.0112      1.0116]
train: WARNING ⚠️ G:\datasets\hand\hand-keypoints\train\images\IMG_00002451.jpg: ignoring corrupt image/label: non-normalized or out of bounds coordinates [     1.0121]
train: WARNING ⚠️ G:\datasets\hand\hand-keypoints\train\images\IMG_00002454.jpg: ignoring corrupt image/label: non-normalized or out of bounds coordinates [      1.023      1.0012]
Many of the callout coordinates are beyond 1

**Reply by glenn-jocher** [reply]
@redsuning the warnings indicate that some images have non-normalized or out-of-bounds coordinates. Please check your dataset annotations to ensure all keypoints are within the expected range. You might need to normalize the coordinates to fit within the [0, 1] range. For more details, you can refer to the dataset documentation.

**Reply by redsuning** [reply]
I know the reason for this warning, but the dataset I'm using is the default Hand-keypoints dataset you uploaded, and I mean your Hand-keypoints dataset labeling data needs to be modified.

**Reply by glenn-jocher** [reply]
Thank you for bringing this to our attention. We'll review the dataset annotations for any inconsistencies. If you have further questions, feel free to ask here.

**Reply by AlgosDuCouteau** [reply]
seem like the issue has not been fixed, i am still get the errors, will it affect training results?

**Reply by glenn-jocher** [reply]
The warnings won't affect training results if the majority of your dataset is intact. We'll review the dataset annotations for any issues. For further details, please refer to the [hand-keypoints documentation](https://docs.ultralytics.com/datasets/pose/hand-keypoints/).

**Reply by berlintop** [reply]
In fact, the official dataset has 5% incorrect labels.

**Reply by glenn-jocher** [reply]
Thank you for pointing this out. We will investigate and address any labeling issues in the Hand Keypoints dataset as soon as possible. Meanwhile, you may exclude problematic samples or adjust the annotations for your use case to ensure optimal training results.

**Comment by UltralyticsAssistant** [comment]
👋 Hello, thank you for your interest in Ultralytics 🚀! It looks like you're exploring hand keypoints estimation datasets for advanced pose estimation. You can find detailed information and resources in our [documentation](https://docs.ultralytics.com/datasets/pose/hand-keypoints/).

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us debug it. An Ultralytics engineer will also assist you soon.

For custom training ❓ questions, please share your dataset, any relevant training logs, and ensure you're following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join our lively community discussions:
- Engage in real-time on [Discord](https://ultralytics.com/discord) 🎧
- Dive into detailed conversations on [Discourse](https://community.ultralytics.com)
- Share insights on our [Subreddit](https://reddit.com/r/ultralytics)

## Upgrade

Make sure you're using the latest version of `ultralytics` by upgrading your package:

```bash
pip install -U ultralytics
```

## Environments

Run YOLOv8 in any of these supported environments with preinstalled dependencies:

- **Notebooks** with a free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM: [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI: [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**: [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

If the badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are currently passing, indicating correct operation on macOS, Windows, and Ubuntu across all modes and tasks.

**Comment by chengziNaomi** [comment]
In yolov8-pose, what effect does the number of key points have on the detection results? Less key points, better results or more key points, better results? Now I want to identify the basket in the field, do I identify the four corner points on the top of the basket or the eight corner points on the basket?

**Reply by glenn-jocher** [reply]
@chengziNaomi the number of keypoints can affect detection results based on your specific application. For identifying a basket, consider the complexity and precision needed. Using more keypoints might provide better detail, but it depends on your use case. For further guidance, you can explore the hand keypoints documentation at [Ultralytics Docs](https://docs.ultralytics.com/datasets/pose/hand-keypoints/).

**Reply by chengziNaomi** [reply]
Thank you for your reply. I have one more question. Why is the ratio of train and valid about 2:1 in this hand
 keypoints datasets? but in seg, the ratio is 10:1:1？Is it a special requirements from yolov8-pose？

**Reply by glenn-jocher** [reply]
The dataset split ratios are often determined by the specific requirements of the task and dataset characteristics. For hand keypoints, a 2:1 ratio might balance training and validation needs effectively, while segmentation tasks might use a 10:1:1 ratio to ensure sufficient data for training, validation, and testing. These choices are not specific to YOLOv8-pose but are tailored to optimize model performance for each dataset type.

**Reply by chengziNaomi** [reply]
Thank you very much! I want to add negative samples to the datasets, what is the way for yolov8-pose to add negative samples?All I need to do is add the background image and blank txt text to train and val，right？

**Reply by glenn-jocher** [reply]
To add negative samples in YOLOv8-pose, you can include background images with corresponding blank `.txt` files in your dataset. This approach helps the model learn to distinguish between relevant and irrelevant features.

**Comment by chrismuntean** [comment]
# I trained a model using this tutorial so you don't have to!

### Feel free to use it as you wish:
### https://github.com/chrismuntean/YOLO11n-pose-hands

**Reply by glenn-jocher** [reply]
@chrismuntean thank you for sharing your model! It's great to see community contributions that enhance the use of YOLO11 for pose estimation tasks.

**Reply by zlj127** [reply]
Wow! Thank you so much for your contribution, bro. It has saved me most of the time for pre-training. You're my god!

**Reply by glenn-jocher** [reply]
Thank you for your kind words! 😊 We're thrilled to hear the Hand Keypoints dataset helped streamline your workflow. For those interested in exploring this dataset further, you can find full training details and annotations in our [Hand Keypoints Dataset documentation](https://docs.ultralytics.com/datasets/pose/hand-keypoints/). The YOLO community always appreciates shared implementations like yours – keep innovating!

**Comment by berlintop** [comment]
In fact, the official dataset has 5% incorrect labels.

**Reply by glenn-jocher** [reply]
Thank you for pointing that out. If the dataset contains mislabeled annotations, you may need to inspect and correct them before training to ensure optimal performance. For further details, you can review the dataset configuration and annotations provided in the [hand-keypoints.yaml](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/cfg/datasets/hand-keypoints.yaml). Let us know if you encounter any specific issues during training.

**Comment by jugal-sheth** [comment]
```
from ultralytics import YOLO

# Load a model
model = YOLO("yolo11n-pose.pt")  # load a pretrained model (recommended for training)

# Train the model
results = model.train(data="hand-keypoints.yaml", epochs=100, imgsz=640) 
```

# Error
train: WARNING ⚠️ D:\projects\hand-track\datasets\dataset\train\images\IMG_00035498.jpg: ignoring corrupt image/label: cannot reshape array of size 67 into shape (2)
train: WARNING ⚠️ D:\projects\hand-track\datasets\dataset\train\images\IMG_00035499.jpg: ignoring corrupt image/label: cannot reshape array of size 67 into shape (2)
train: WARNING ⚠️ D:\projects\hand-track\datasets\dataset\train\images\IMG_00035500.jpg: ignoring corrupt image/label: cannot reshape array of size 67 into shape (2)
.
.
.
.
.
.
.
train: WARNING ⚠️ D:\projects\hand-track\datasets\dataset\train\images\IMG_00035501.jpg: ignoring corrupt image/label: cannot reshape array of size 67 into shape (2)
train: WARNING ⚠️ D:\projects\hand-track\datasets\dataset\train\images\IMG_00035503.jpg: ignoring corrupt image/label: cannot reshape array of size 67 into shape (2)
train: WARNING ⚠️ D:\projects\hand-track\datasets\dataset\train\images\IMG_00035504.jpg: ignoring corrupt image/label: cannot reshape array of size 67 into shape (2)
train: WARNING ⚠️ D:\projects\hand-track\datasets\dataset\train\images\IMG_00035505.jpg: ignoring corrupt image/label: cannot reshape array of size 67 into shape (2)
train: WARNING ⚠️ D:\projects\hand-track\datasets\dataset\train\images\IMG_00035507.jpg: ignoring corrupt image/label: cannot reshape array of size 67 into shape (2)
train: WARNING ⚠️ D:\projects\hand-track\datasets\dataset\train\images\IMG_00035509.jpg: ignoring corrupt image/label: cannot reshape array of size 67 into shape (2)
WARNING ⚠️ No images found in D:\projects\hand-track\datasets\dataset\train\labels.cache, training may not work correctly. See https://docs.ultralytics.com/datasets for dataset formattingg guidance.
Traceback (most recent call last):
  File "d:\projects\hand-track\test.py", line 36, in <module>
    model.train(data="config.yaml", epochs=100, imgsz=896, device=0)
  File "C:\Users\jugal\miniconda3\envs\hand\Lib\site-packages\ultralytics\engine\model.py", line 808, in train
    self.trainer.train()
  File "C:\Users\jugal\miniconda3\envs\hand\Lib\site-packages\ultralytics\engine\trainer.py", line 207, in train
    self._do_train(world_size)
  File "C:\Users\jugal\miniconda3\envs\hand\Lib\site-packages\ultralytics\engine\trainer.py", line 322, in _do_train
    self._setup_train(world_size)
  File "C:\Users\jugal\miniconda3\envs\hand\Lib\site-packages\ultralytics\engine\trainer.py", line 286, in _setup_train
    self.train_loader = self.get_dataloader(self.trainset, batch_size=batch_size, rank=LOCAL_RANK, mode="train")
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\jugal\miniconda3\envs\hand\Lib\site-packages\ultralytics\models\yolo\detect\train.py", line 49, in get_dataloader
    dataset = self.build_dataset(dataset_path, mode, batch_size)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\jugal\miniconda3\envs\hand\Lib\site-packages\ultralytics\models\yolo\detect\train.py", line 43, in build_dataset
    return build_yolo_dataset(self.args, img_path, batch, self.data, mode=mode, rect=mode == "val", stride=gs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\jugal\miniconda3\envs\hand\Lib\site-packages\ultralytics\data\build.py", line 99, in build_yolo_dataset
    return dataset(
           ^^^^^^^^
  File "C:\Users\jugal\miniconda3\envs\hand\Lib\site-packages\ultralytics\data\dataset.py", line 64, in __init__
    super().__init__(*args, **kwargs)
  File "C:\Users\jugal\miniconda3\envs\hand\Lib\site-packages\ultralytics\data\base.py", line 74, in __init__
    self.labels = self.get_labels()
                  ^^^^^^^^^^^^^^^^^
  File "C:\Users\jugal\miniconda3\envs\hand\Lib\site-packages\ultralytics\data\dataset.py", line 161, in get_labels
    len_cls, len_boxes, len_segments = (sum(x) for x in zip(*lengths))
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: not enough values to unpack (expected 3, got 0)

**Reply by glenn-jocher** [reply]
@jugal-sheth hi there! It looks like your labels might be incorrectly formatted for hand keypoints estimation. The hand-keypoints dataset requires 21 keypoints with 3 dimensions each (x,y,visibility), but your labels appear to have 67 values instead of the required 63 (21*3). Please verify your label files match the required format shown in our hand-keypoints documentation: https://docs.ultralytics.com/datasets/pose/hand-keypoints/#dataset-yaml 💡

For commercial licensing options, visit https://ultralytics.com/license. Let us know if you need any other assistance!

**Reply by jugal-sheth** [reply]
Hi @glenn-jocher Thank you for your reply
I haven't changed anything in sample code given on this website
https://docs.ultralytics.com/datasets/pose/hand-keypoints/#how-do-i-train-a-yolo11-model-on-the-hand-keypoints-dataset

**Reply by glenn-jocher** [reply]
@jugal-sheth hi there! It seems your custom labels may not include the required bounding box coordinates alongside keypoints. The YOLO pose format requires each label to start with class, x_center, y_center, width, height (5 values) followed by 21 keypoints (63 values), totaling 68 values per label. Your 67-value labels suggest missing one element. Please verify your label format matches our [pose estimation requirements](https://docs.ultralytics.com/datasets/pose/#ultralytics-yolo-format) exactly. Let us know if you need further clarification! 🚀

**Comment by changshen123** [comment]
Hi @glenn-jocher  i want to konw that how much epochs are suitable for your dataset？

**Comment by samasadii** [comment]
I've noticed in this dataset almost all of the points are labeled as visible, even the ones that are not visible in the image.

**Comment by RedAbraham** [comment]
what's the meaning of output(1*300*69)? I wanna deploy the model by myself but I don't know how to analyze the output

**Reply by glenn-jocher** [reply]
That `(1, 300, 69)` is the exported pose head output shaped as `(batch, num_predictions, num_channels)`; for Hand Keypoints `kpt_shape: [21, 3]`, `69 = 4 (box) + 1 (objectness) + 1 (class score, since `names: {0: hand}`) + 21*3 (keypoints x,y,visibility)`—see the dataset definition in the [Hand Keypoints dataset docs](https://docs.ultralytics.com/datasets/pose/hand-keypoints/) and the general [pose dataset/output format overview](https://docs.ultralytics.com/datasets/pose/).  

```python
# out: (1, 300, 69)
pred = out[0]                  # (300, 69)
boxes_xywh = pred[:, 0:4]      # x,y,w,h (normalized)
obj = pred[:, 4]               # objectness
cls = pred[:, 5:6]             # class score (1 class)
kpts = pred[:, 6:].reshape(-1, 21, 3)  # (300, 21, 3) -> x,y,vis (normalized)
```

If you tell me your export format (e.g. ONNX/TensorRT) and whether you exported with `nms=True` (raw vs post-NMS outputs), I can confirm the exact channel order for your case—feel free to drop the tensor names and a small snippet of the output here in this discussion so the team can see it.

**Comment by vbalab** [comment]
# Ready-to-use YOLO26s-pose fine-tuned for hand keypoints!

#### The link: https://github.com/vbalab/YOLO26s-pose-hands

**Reply by glenn-jocher** [reply]
Thanks for sharing your [YOLO26s hand-keypoints weights repo](https://github.com/vbalab/YOLO26s-pose-hands) — for anyone trying it, please include (in that repo) the exact `ultralytics` version + training args and a validation result on the official [Hand Keypoints dataset](https://docs.ultralytics.com/datasets/pose/hand-keypoints/) so others can reproduce/compare easily; quick sanity-check command:

```bash
yolo pose val model=path/to/best.pt data=hand-keypoints.yaml imgsz=640
```
